	// ==IMPORT      ==================================================================================
	import { Request, Response, NextFunction } from "express";
	import { ZodError } from "zod";
	import { logConsole } from "./logger.js";

	// == TYPE, ENUM, INTERFACE  =======================================================================
	export enum ApiCodeErreur {
		parsing= "PARSING SQL",
		param = "PARAMETRES",
		contrainte = "CONTRAINTE",
		inconnu = "INCONNU",
		bdReference = "BASE DE REFERENCE",
		reqInvalide = "REQUETE INVALIDE",
		urlInconnue = "URL INCONNUE",
		authentification = "ERREUR AUTHENTIFICATION"
	}
	
	export interface AppError {
		code: string;
		message: string;
		details?: string[];
	}

	export interface Resultat {
		success: boolean;
		message?: string;
		erreur?: AppError;
		donnees?: any 
	}

	export class ContraintBaseReferenceError extends Error {
	statusCode: number;

	constructor(message: string) {
		super(message);
		this.name = 'ContraintBaseReferenceError';
		this.statusCode = 400;
	}
	}

	// HttpException.ts
	export class HttpException extends Error {
		public status: number;
		public message: string;
		
		constructor(status: number, message: string) {
			super(message);
			this.status = status;
			this.message = message;
		}
	}

	// Classe d'erreur personnalis√©e
	export class CustomAppException extends Error implements Resultat {
		public success: boolean;
		public erreur?: AppError;
	
		constructor(resultat: Resultat) {
		super(resultat.message); // Appelle le constructeur de la classe Error avec le message
		this.name = 'CustomAppException'; // Nom de l'erreur pour un meilleur d√©bogage
		this.success = resultat.success;
		this.erreur = resultat.erreur;
		
		// Le code ci-dessous est important pour maintenir une trace de pile (stack trace)
		// correcte, surtout dans les environnements comme Node.js et les navigateurs modernes.
		if (Error.captureStackTrace) {
			Error.captureStackTrace(this, CustomAppException);
		}
		}
	}

	// == DECLARATION     ==============================================================================
	const emoji = "üçî‚Äã";
	const viewLog = true;
	const module = 'ww-package/utilitaires/error';

	// == FONCTION        ==============================================================================

	// Fonction de garde de type pour v√©rifier si un objet est de type AppError
	export function isAppError(obj: any): obj is AppError {
		// V√©rifie si l'objet n'est pas nul et est bien un objet
		if (typeof obj !== 'object' || obj === null) {
		return false;
		}
		
		// V√©rifie si les propri√©t√©s 'code' et 'message' sont pr√©sentes et de type string
		const hasCodeAndMessage = 'code' in obj && typeof obj.code === 'string' &&
										'message' in obj && typeof obj.message === 'string';
	
		// V√©rifie si la propri√©t√© 'details' est optionnelle et de type string[] si elle existe
		const hasOptionalDetails = !('details' in obj) || 
											(Array.isArray(obj.details) && 
											obj.details.every((detail: string) => typeof detail === 'string')); // <--- CORRECTION ICI
	
		return hasCodeAndMessage && hasOptionalDetails;
	}
	
	// Fonction format AppError
	export function formatAppError(error: AppError, useHtml = false): string {
		const separator = useHtml ? '<br>' : '\n';
		const bullet = useHtml ? '‚Ä¢ ' : '- '; // Vous pouvez utiliser '‚Ä¢ ' pour un point ou '- ' pour un tiret
	
		let result = error.message;
	
		if (error.details && error.details.length > 0) {
		// Pr√©fixe chaque d√©tail avec la puce avant de les joindre
		const detailsWithBullets = error.details.map(detail => `${bullet}${detail}`);
		const detailsString = detailsWithBullets.join(separator);
		
		result += `${separator}${detailsString}`;
		}
	
		return result;
	}
	
	// Fonction handleError
	export function handleError(err: unknown, req: Request, res: Response, next: NextFunction) {
	logConsole(viewLog, emoji, module + '/handleError', 'D√©but', '-');

		// ‚úÖ Cas 0 : Erreur HttpException personnalis√©e
		if (err instanceof HttpException) {
			logConsole(viewLog, emoji, module + '/handleError', 'Erreur HttpException : err', err);
			return res.status(err.status).json({
				code: ApiCodeErreur.authentification,
				message: err.message
			});	
		}
		
		// ‚úÖ Cas 1 : Erreur de parsing JSON
		if (err instanceof SyntaxError && 'status' in err && (err as any).status === 400 && 'body' in err) {
			const erreurs: AppError = {
				code: ApiCodeErreur.parsing,
				message: "Erreur de syntaxe JSON dans la requ√™te.",
				details: [err.message]      
			}
			return res.status(400).json(erreurs);
		} 
	
		// ‚úÖ Cas 2 : Erreur validation Zod
		if (err instanceof ZodError) {
			logConsole(viewLog, emoji, module + '/handleError', 'Erreur de validation Zod : err', err);
			const details:string[]= err.issues.map(e => (e.message));

			const erreurs: AppError = {
				code: ApiCodeErreur.param,
				message: "Param√®tres invalides",
				details: details
			}

			return res.status(400).json(erreurs);
		}
		
		// ‚úÖ Cas 3 : Erreur de validation des donn√©es du dictionnaire de r√©f√©rence
		if (err instanceof ContraintBaseReferenceError) {
			const erreurs: AppError = {
				code: ApiCodeErreur.bdReference,
				message: "Erreur dans les donn√©es fournies du dictionnaire de r√©f√©rence",
				details: [err.message]
			}
			return res.status(err.statusCode).json(erreurs);
		}

	// ‚úÖ Cas 4 : Erreurs generique
		if (err instanceof Error) {	
			logConsole(viewLog, emoji, module + '/handleError',"Cas ou l'erreur est de type ERROR",err.message);

			let status = (err as any)?.status;
	
			// Si status est undefined, null, 0 ou autre "falsy", on force √† 500
			if (!status || typeof status !== 'number' || status < 100 || status > 599) {
			status = 500;
			}
			logConsole(viewLog, emoji, module + '/handleError',"- statut",status);
			
			// Erreur constrainte SQL
			if (status === 'SQLITE_CONSTRAINT') {
				return res.status(409).json({
					code: ApiCodeErreur.contrainte,
					message: "Conflit : cet √©l√©ment existe d√©j√†."
				});
			}
	
			if (status === 400) {
				const erreurs: AppError = {
					code: ApiCodeErreur.reqInvalide,
					message: "Requ√™te invalide.",
					details: [err.message]
				}
			return res.status(400).json(erreurs);
			}

			if (status === 401) {
				const erreurs: AppError = {
					code: ApiCodeErreur.authentification,
					message: "Erreur lors de la v√©rification de l'authentification",
					details: [err.message]
				}
				return res.status(401).json(erreurs);	
			}

			if (status === 404) {
				const erreurs: AppError = {
					code: ApiCodeErreur.urlInconnue,
					message: "URL Inconnue.",
					details: [err.message]
				}
			return res.status(404).json(erreurs);
			}
	
			const erreurs: AppError = {
				code: ApiCodeErreur.inconnu,
				message: "Erreur interne du serveur.",
				details: [err.message]
			}
			return res.status(status).json(erreurs);
			
		}
		
	// ‚úÖ Cas 5 : Erreur g√©n√©rique non pr√©vue
		logConsole(viewLog, emoji, module + '/handleError', 'Erreur non pr√©vue', err);
		const erreurs: AppError = {
			code: ApiCodeErreur.inconnu,
			message: "Erreur interne du serveur.",
			details: [String(err)]
		}
	return res.status(500).json(erreurs);

	} 
